### ISSUE TITLE
Refactor file search mechanism

### ISSUE THREAD

**Issue by ewels** [ewels/MultiQC#202]
It would be good to rewrite how the log file discovery works, in preparation for the restructuring discussed in #181. Currently, the core script first looks through the analysis directories and collects the relevant files (ignoring big files / those matching exclusion patterns etc). These are then passed on to each module, which then searches through them for logs.

What we could like to do instead is for the core script to do the searching for all modules on that first pass. A list of matching files can then be passed on to each module. This first search would go by filename first, then content, and files matched from one module could then be ignored by searches for the next (unless specified otherwise). Hopefully this will speed up execution time quite a bit, especially when there are a lot of samples. It should also pave the way for future changes if/when we remove modules from the core tool.

**Comment by ewels** [comment]
Would be good to add in an option to search by regex. Also an option for regex + string to only search first `n` lines.

**Comment by ewels** [comment]
Currently, the 5MB filesize limit applies to files matched by filename too. Maybe worth removing this restriction.

**Comment by ewels** [comment]
Ok, _definitely_ going to do this for the next release..!

**Comment by ewels** [comment]
If it doesn't already work, would be nice to be able to search by _both_ filename and content. For example, search all `*.txt` files for a string.

**Comment by ewels** [comment]
Now (v1.0 release) may be as good a time as ever to make the breaking-change of forcing single-level search patterns. The current mixture is confusing and problematic. For example, if _only_ changing one Picard pattern, then _all_ Picard patterns must be specified in the user config, otherwise MultiQC breaks.

Down side is that this will force many people to change their configs. Perhaps could use underscores and try to find the new pattern in case of an Exception?

**Comment by epruesse** [comment]
You use `dict.update()` to merge the yaml config files, which is non-recursive. Making it recursive is simple enough though (loop over `iteritems()`, recurse if `instanceof(item, collections.Mapping)` encountered. A recursive dictionary update function probably just doesn't exist in `dict` because there are so many variants of what you may want to do (e.g. do you want to replace or append in case a list object is encountered?)

**Comment by ewels** [comment]
A recursive approach like this is certainly possible, but it's perhaps not super trivial. There are lots of different possibilities for the search patterns (with more to come) - for instance, lists of file search strings; dicts with a content search and a filename search; dicts containing multiple configuration keys (eg. number of lines to search and so on). It still wouldn't be very difficult, but it would have to rule out a bunch of cases before recursing. It would also make life more difficult when accessing the test results - a simple key lookup is less complicated than having multiple depths of results.

It's a shame that I wasn't more strict with myself at the start and wrote it like this in the first place!

Phil

**Comment by epruesse** [comment]
I've hacked up something yesterday and will open a PR for you to review it later today.

**Comment by epruesse** [comment]
There are still some caveats to fix, but I didn't want to invest enormous amounts of time into sth you may not like.

The concept is to use `file` for the content detection. `file(1)` is configured using the `magic(5)` file format, has been around for ages, should be flexible enough to encode pretty much everything and is (being a classic unix tool) fast.

For compatability, I'm converting the searchpatterns.yaml style config on the fly into a magic file. For more flexibility, writing those directly and bundling them with each module would be an option, or extending the parser to create them from more complex rules.

Briefly, the magic files work like this:

Each line is a rule comprising four words: offset, type, value and message. The offset specifies how far into the file the rule should be checked, the type what kind of match it will be, the value is the comparison value and the message is what's printed on match success.

Useful match types are "search" and "regex", with "search/N" searching the first N bytes for the match string and "regex/40l" searching the first 40 lines.

Rules can be chained by prefixing them with ">" in a "indentation hierarchy" fashion. Rules are only executed if the super-rule succeeded. So "and" combinations go down the tree and "or" combinations are siblings on a tree node. Adding the pseudo rule types "name" and "use" which allow defining and calling trees, that pretty much gives all the power needed. It'd suffice to detect the phred offset in fq.gz files (file can unzip on the fly).

**Comment by ewels** [comment]
This issue has been addressed in PR #449. I have flattened the config keys [as suggested above](#issuecomment-289899903), but used the slash syntax (`module/section`) used by @epruesse in your PR. I was thinking that this could be used to add backwards-compatibility for people with the old-style syntax in their custom config files, though I haven't written this in yet.

As far as I can tell, my code does all of the stuff mentioned above, but using pure Python. There are now four methods of searching - filename glob, filename regex, string in contents, regex in contents. Searches can then also be limited by filesize, number of lines into the file for each search. Filename and contents can be combined in a single YAML dict for `AND` logic, or in a list for `OR` logic (or both, for a combination). It doesn't do unzipping on the fly, but I can't think of any cases where we need to do that (currently). I'm sure that it's slower than using `file`, but by how much I'm not sure.

Phil
